File: scripts/scrape_cn.py
"""Scrape official China (国服) Wild Rift champion win/pick/ban rates.

Source: Tencent's official Wild Rift site (lolm.qq.com) exposes daily champion
statistics via two JSON endpoints used by its 国服数据 page:

  hero list:  https://game.gtimg.cn/images/lgamem/act/lrlib/js/heroList/hero_list.js
  rank data:  https://mlol.qt.qq.com/go/lgame_battle_info/hero_rank_list_v2

The rank data is keyed by cumulative rank bracket (0-4: Diamond+ / Master+ /
Challenger+ / Sovereign) then by position (1-5: top/jungle/mid/bot/support). Each entry has
win_rate_percent, appear_rate_percent (pick), forbid_rate_percent (ban) and
strength_level (tier). This is official, daily, high-elo data — a real CN source
to sit alongside our EU top-50 dataset.

Champion names are Chinese; we bridge them to our English roster once (via Gemini,
validated against site.json) and cache the map to data/cn_hero_map.json.

Output: data/cn_winrates.json

Run (Gemini key only needed the first time, to build the name map):
    GEMINI_API_KEY=... python -m scripts.scrape_cn
"""
from __future__ import annotations

import json
import re
from pathlib import Path

import requests

ROOT = Path(__file__).resolve().parent.parent
SITE = ROOT / "web-next" / "src" / "data" / "site.json"
NAME_MAP = ROOT / "data" / "cn_hero_map.json"
OUT = ROOT / "data" / "cn_winrates.json"
WEB_OUT = ROOT / "web-next" / "src" / "data" / "cn.json"  # what the frontend reads

HERO_LIST_URL = "https://game.gtimg.cn/images/lgamem/act/lrlib/js/heroList/hero_list.js"
RANK_URL = "https://mlol.qt.qq.com/go/lgame_battle_info/hero_rank_list_v2"
HEADERS = {"User-Agent": "Mozilla/5.0", "Referer": "https://lolm.qq.com/"}

# bracket keys verified against the site's rank tabs (钻石/大师/王者/峡谷之巅)
BRACKET_LABELS = {
    "0": "All ranks", "1": "Diamond+", "2": "Master+", "3": "Challenger+", "4": "Sovereign",
}
# Challenger and above (includes Sovereign): a larger, steadier top-elo sample
# than the Sovereign tier alone, so it's what the frontend surfaces as "highest".
DEFAULT_BRACKET = "3"
# position codes verified empirically against the champion pools per lane
POSITION_LABELS = {"1": "Mid", "2": "Baron", "3": "Dragon", "4": "Support", "5": "Jungle"}


def _slug(name: str) -> str:
    s = name.lower().replace("&", "and").replace("'", "")
    return re.sub(r"[^a-z0-9]+", "-", s).strip("-")


def fetch_hero_list() -> dict[str, str]:
    """heroId -> Chinese name."""
    txt = requests.get(HERO_LIST_URL, headers=HEADERS, timeout=20).text
    obj = json.loads(txt)
    return {hid: h["name"] for hid, h in obj["heroList"].items()}


def fetch_rank() -> dict:
    r = requests.get(RANK_URL, headers=HEADERS, timeout=20).json()
    return r["data"]


def build_name_map(cn_names: dict[str, str], english: list[str]) -> dict[str, str]:
    """heroId -> English name. Cached; uses Gemini once to translate CN->EN."""
    if NAME_MAP.exists():
        return json.loads(NAME_MAP.read_text(encoding="utf-8"))

    from google import genai
    from google.genai import types

    prompt = (
        "Map each Chinese Wild Rift champion name to its exact English name, chosen "
        "ONLY from the provided English list. Return a JSON object {chineseName: englishName}. "
        "If a Chinese name has no match in the list, map it to \"\".\n\n"
        f"English champion names:\n{json.dumps(english, ensure_ascii=False)}\n\n"
        f"Chinese names to map:\n{json.dumps(sorted(set(cn_names.values())), ensure_ascii=False)}"
    )
    client = genai.Client()
    resp = client.models.generate_content(
        model="gemini-2.5-flash", contents=prompt,
        config=types.GenerateContentConfig(response_mime_type="application/json", temperature=0),
    )
    cn_to_en = json.loads(re.search(r"\{.*\}", resp.text, re.DOTALL).group(0))
    eng_set = set(english)
    hero_map: dict[str, str] = {}
    unmatched = []
    for hid, cn in cn_names.items():
        en = cn_to_en.get(cn, "")
        if en and en in eng_set:
            hero_map[hid] = en
        else:
            unmatched.append(f"{hid}:{cn}")
    NAME_MAP.write_text(json.dumps(hero_map, ensure_ascii=False, indent=2), encoding="utf-8")
    if unmatched:
        print(f"  ! {len(unmatched)} unmapped heroes: {unmatched}")
    print(f"  name map: {len(hero_map)}/{len(cn_names)} bridged -> {NAME_MAP.relative_to(ROOT)}")
    return hero_map


def main() -> None:
    site_champs = json.loads(SITE.read_text(encoding="utf-8"))["champions"]
    english = [c["name"] for c in site_champs]
    eu_role = {c["name"]: c["role"] for c in site_champs}  # align CN lane to EU role
    cn_names = fetch_hero_list()
    rank = fetch_rank()
    hero_map = build_name_map(cn_names, english)

    date = None
    # per champion: byBracket -> {wr, pick, ban, strength, position} at primary lane
    champs: dict[str, dict] = {}
    for bracket, positions in rank.items():
        for pos, entries in positions.items():
            for e in entries:
                hid = e["hero_id"]
                en = hero_map.get(hid)
                if not en:
                    continue
                date = date or e.get("dtstatdate")
                rec = champs.setdefault(en, {
                    "name": en, "slug": _slug(en), "heroId": hid,
                    "cnName": cn_names.get(hid, ""), "byBracket": {},
                })
                pick = float(e["appear_rate_percent"])
                cand = {
                    "winRate": float(e["win_rate_percent"]),
                    "pickRate": pick,
                    "banRate": float(e["forbid_rate_percent"]),
                    "strength": int(e["strength_level"]),
                    "position": POSITION_LABELS.get(pos, pos),
                }
                # Align the CN lane to the champion's EU role so cross-server stats
                # compare like-for-like (e.g. Olaf is Baron on EU but most-picked
                # Jungle in CN — we want his Baron win rate). The EU-role lane always
                # wins over a more-picked lane; ties fall back to highest pick rate.
                target = eu_role.get(en)
                stored = rec["byBracket"].get(bracket)
                cand_match = cand["position"] == target
                if stored is None:
                    take = True
                elif cand_match != (stored["position"] == target):
                    take = cand_match
                else:
                    take = pick > stored["pickRate"]
                if take:
                    rec["byBracket"][bracket] = cand

    out = {
        "source": "lolm.qq.com (official China server data)",
        "date": date,
        "bracketLabels": BRACKET_LABELS,
        "defaultBracket": DEFAULT_BRACKET,
        "nChampions": len(champs),
        "champions": sorted(champs.values(), key=lambda c: c["name"]),
    }
    payload = json.dumps(out, ensure_ascii=False, indent=2)
    OUT.write_text(payload, encoding="utf-8")
    WEB_OUT.write_text(payload, encoding="utf-8")  # keep the frontend copy in sync
    print(f"wrote {OUT.relative_to(ROOT)} + {WEB_OUT.relative_to(ROOT)} ({len(champs)} champions, date {date})")

    # dated snapshot (highest bracket) for patch-over-patch history
    from datetime import datetime as _dt
    snap_date = (_dt.strptime(date, "%Y%m%d").date().isoformat()
                 if date and len(date) == 8 and date.isdigit() else (date or "unknown"))
    hist = ROOT / "data" / "history" / "cn"
    hist.mkdir(parents=True, exist_ok=True)
    snap_champs = {}
    for c in champs.values():
        e = c["byBracket"].get(DEFAULT_BRACKET)
        if e:
            snap_champs[c["slug"]] = {"wr": e["winRate"], "pick": e["pickRate"], "ban": e["banRate"]}
    (hist / f"{snap_date}.json").write_text(
        json.dumps({"date": snap_date, "region": "cn", "champions": snap_champs},
                   ensure_ascii=False, indent=2), encoding="utf-8")
    print(f"  snapshot: data/history/cn/{snap_date}.json ({len(snap_champs)} champions)")


if __name__ == "__main__":
    main()

File: scripts/scrape_cn.mjs
/** Scrape official China Wild Rift win/pick/ban rates from Tencent. */
import { mkdir, readFile, writeFile } from "node:fs/promises";
import path from "node:path";

const ROOT = process.cwd();
const SITE = path.join(ROOT, "web-next", "src", "data", "site.json");
const NEW_CHAMPIONS = path.join(ROOT, "web-next", "src", "data", "new_champions.json");
const NAME_MAP = path.join(ROOT, "data", "cn_hero_map.json");
const OUT = path.join(ROOT, "data", "cn_winrates.json");
const WEB_OUT = path.join(ROOT, "web-next", "src", "data", "cn.json");
const PREVIOUS_OUT = path.join(ROOT, "data", "cn_winrates_prev.json");
const MOVERS_OUT = path.join(ROOT, "web-next", "src", "data", "cn_movers.json");
const HISTORY_DIR = path.join(ROOT, "data", "history", "cn");

// PINNED BASELINE for the home page's patch movers.
//
// The default bracket used to be compared against yesterday's scrape, which
// made "Patch 7.2c Biggest Winners & Losers" a one-day drift measurement --
// both ends sat days after the patch and it said nothing about 7.2c. This pins
// it to the last pre-patch snapshot instead, so the window widens by a day with
// every daily run and always measures the patch.
//
// 7.2c went live 2026-08-12 09:00 UTC and this job runs at 06:00 UTC, so the
// 12th snapshot was taken three hours BEFORE the patch landed: a genuinely
// clean baseline. Update this the next time a patch ships.
const MOVERS_BASELINE_DATE = "2026-08-12";
// History snapshots only carry the DEFAULT bracket, so only that bracket can be
// pinned. The rest keep the day-over-day comparison, and every bracket now
// reports which baseline it used so the UI cannot mislabel them.

const HERO_LIST_URL = "https://game.gtimg.cn/images/lgamem/act/lrlib/js/heroList/hero_list.js";
const RANK_URL = "https://mlol.qt.qq.com/go/lgame_battle_info/hero_rank_list_v2";
const HEADERS = { "User-Agent": "Mozilla/5.0", Referer: "https://lolm.qq.com/" };
const BRACKET_LABELS = {
  "1": "Diamond+",
  "2": "Master+",
  "3": "Challenger",
  "4": "Legendary",
};
const DEFAULT_BRACKET = "3";
const POSITION_LABELS = { "1": "Mid", "2": "Baron", "3": "Dragon", "4": "Support", "5": "Jungle" };

const jsonFile = async (file) => JSON.parse(await readFile(file, "utf8"));
const slug = (name) => name.toLowerCase().replaceAll("&", "and").replaceAll("'", "").replace(/[^a-z0-9]+/g, "-").replace(/^-|-$/g, "");

async function fetchJson(url) {
  const response = await fetch(url, { headers: HEADERS });
  if (!response.ok) throw new Error(`${url} returned ${response.status}`);
  return response.json();
}

async function main() {
  const site = await jsonFile(SITE);
  const newChampions = await jsonFile(NEW_CHAMPIONS);
  const heroMap = await jsonFile(NAME_MAP);
  const checkedInPrevious = await jsonFile(OUT).catch(() => null);
  const euRole = new Map(
    [...site.champions, ...newChampions.champions].map((champion) => [champion.name, champion.role]),
  );
  const heroPayload = await fetchJson(HERO_LIST_URL);
  const cnNames = Object.fromEntries(Object.entries(heroPayload.heroList).map(([id, hero]) => [id, hero.name]));
  const rankPayload = await fetchJson(RANK_URL);
  const rank = rankPayload.data;
  if (!rank || typeof rank !== "object") throw new Error("Tencent rank response did not contain data");

  let date = null;
  const champions = new Map();
  const unmapped = new Set();

  for (const [bracket, positions] of Object.entries(rank)) {
    // Tencent's public page only exposes keys 1-4. Key 0 is legacy/internal
    // data and is explicitly discarded by the official frontend.
    if (!BRACKET_LABELS[bracket]) continue;
    for (const [position, entries] of Object.entries(positions)) {
      for (const entry of entries) {
        const heroId = String(entry.hero_id);
        const name = heroMap[heroId];
        if (!name) {
          unmapped.add(`${heroId}:${cnNames[heroId] ?? "unknown"}`);
          continue;
        }
        date ||= entry.dtstatdate || null;
        if (!champions.has(name)) {
          champions.set(name, {
            name,
            slug: slug(name),
            heroId,
            cnName: cnNames[heroId] ?? "",
            byBracket: {},
          });
        }

        const record = champions.get(name);
        const candidate = {
          winRate: Number(entry.win_rate_percent),
          pickRate: Number(entry.appear_rate_percent),
          banRate: Number(entry.forbid_rate_percent),
          strength: Number(entry.strength_level),
          position: POSITION_LABELS[position] ?? position,
        };
        const targetRole = euRole.get(name);
        const stored = record.byBracket[bracket];
        const candidateMatches = candidate.position === targetRole;
        const storedMatches = stored?.position === targetRole;
        const take = !stored
          ? true
          : candidateMatches !== storedMatches
            ? candidateMatches
            : candidate.pickRate > stored.pickRate;
        if (take) {
          record.byBracket[bracket] = candidate;
        }
      }
    }
  }

  if (unmapped.size) console.warn(`Skipped ${unmapped.size} unmapped heroes: ${[...unmapped].join(", ")}`);
  const sorted = [...champions.values()].sort((left, right) => left.name.localeCompare(right.name));
  const output = {
    source: "lolm.qq.com (official China server data)",
    date,
    bracketLabels: BRACKET_LABELS,
    defaultBracket: DEFAULT_BRACKET,
    nChampions: sorted.length,
    champions: sorted,
  };
  const payload = `${JSON.stringify(output, null, 2)}\n`;
  await Promise.all([writeFile(OUT, payload, "utf8"), writeFile(WEB_OUT, payload, "utf8")]);

  // Compare every published bracket with the last checked-in scrape. A
  // same-day rerun falls back to the preserved pre-refresh file so manually
  // rerunning the job does not erase every movement indicator.
  let previous = checkedInPrevious;
  if (!previous || previous.date === date) previous = await jsonFile(PREVIOUS_OUT).catch(() => null);
  const previousBySlug = new Map((previous?.champions ?? []).map((champion) => [champion.slug, champion]));

  // The pinned pre-patch snapshot, for the default bracket only.
  const baseline = await jsonFile(path.join(HISTORY_DIR, `${MOVERS_BASELINE_DATE}.json`))
    .catch(() => null);
  if (!baseline) {
    console.warn(`No pinned baseline at ${MOVERS_BASELINE_DATE}; the default bracket falls back to day-over-day.`);
  }
  const baselineDates = {};
  const moversByBracket = {};
  for (const bracket of Object.keys(BRACKET_LABELS)) {
    // Only the default bracket has a pinned pre-patch baseline available,
    // because history snapshots store that bracket alone.
    const pinned = bracket === DEFAULT_BRACKET && baseline;
    baselineDates[bracket] = pinned ? MOVERS_BASELINE_DATE.replace(/-/g, "") : (previous?.date ?? "");
    moversByBracket[bracket] = sorted.flatMap((champion) => {
      const oldWr = pinned
        ? baseline.champions?.[champion.slug]?.wr
        : previousBySlug.get(champion.slug)?.byBracket?.[bracket]?.winRate;
      const newEntry = champion.byBracket[bracket];
      if (oldWr == null || !newEntry) return [];
      return [{
        slug: champion.slug,
        name: champion.name,
        oldWr,
        newWr: newEntry.winRate,
        delta: Math.round((newEntry.winRate - oldWr) * 100) / 100,
        pickRate: newEntry.pickRate,
      }];
    }).sort((left, right) => right.delta - left.delta);
  }
  const movers = moversByBracket[DEFAULT_BRACKET];
  await writeFile(MOVERS_OUT, `${JSON.stringify({
    // The default bracket's window, which is what the home page shows.
    beforeDate: baselineDates[DEFAULT_BRACKET] || (previous?.date ?? ""),
    afterDate: date,
    // Per-bracket baselines: the default bracket is pinned pre-patch, the rest
    // are day-over-day, and the UI needs to know which is which.
    baselineByBracket: baselineDates,
    patch: "",
    scope: "China · Challenger",
    defaultBracket: DEFAULT_BRACKET,
    bracketLabels: BRACKET_LABELS,
    champions: movers,
    byBracket: moversByBracket,
  }, null, 2)}\n`, "utf8");

  const snapshotDate = /^\d{8}$/.test(date ?? "")
    ? `${date.slice(0, 4)}-${date.slice(4, 6)}-${date.slice(6, 8)}`
    : date || "unknown";
  const historyDir = path.join(ROOT, "data", "history", "cn");
  await mkdir(historyDir, { recursive: true });
  const snapshotChampions = {};
  for (const champion of sorted) {
    const entry = champion.byBracket[DEFAULT_BRACKET];
    if (entry) snapshotChampions[champion.slug] = { wr: entry.winRate, pick: entry.pickRate, ban: entry.banRate };
    // Every bracket, not just the default one. Snapshots used to store the
    // default alone, which is why only that bracket can be pinned to a
    // pre-patch baseline today. Recording all four means the next patch can
    // pin every bracket instead of one.
    const byBracket = {};
    for (const [bracket, value] of Object.entries(champion.byBracket ?? {})) {
      if (value) byBracket[bracket] = { wr: value.winRate, pick: value.pickRate, ban: value.banRate };
    }
    if (Object.keys(byBracket).length && snapshotChampions[champion.slug]) {
      snapshotChampions[champion.slug].byBracket = byBracket;
    }
  }
  await writeFile(
    path.join(historyDir, `${snapshotDate}.json`),
    `${JSON.stringify({ date: snapshotDate, region: "cn", champions: snapshotChampions }, null, 2)}\n`,
    "utf8",
  );
  console.log(`Wrote CN data for ${sorted.length} champions, source date ${date}, ${Object.keys(BRACKET_LABELS).length} published brackets, ${movers.length} Challenger movement rows.`);
}

await main();
